### PR TITLE
vpImageMorphology

### PR DESCRIPTION
**Pull request by rolalaro**
vpImageMorphology::dilatation(vpImage&, vpConnexityType) and vpImageMorphology::erosion(vpImage&, vpConnexityType) are now templated to compute reciproquely the max and min of the input image on a 3x3 grid.

### CODE REVIEW

**Comment by fspindle** [comment]
@rolalaro Thanks, but I'm afraid this PR will break compatibility with previous releases.

Is there any way of adding non-template functions as deprecated with something like (not tested):

```
#if defined(VISP_BUILD_DEPRECATED_FUNCTIONS)
  /*!
    @name Deprecated functions
  */
  //@{
  /*!
     \deprecated Provided only for compat with previous releases. Use rather the template function erosion(vpImage<T> &, const vpConnexityType &)
     
     Add here the description of the function...
   */
  vp_deprecated static void erosion(vpImage<unsigned char> &I, const vpConnexityType &connexity = CONNEXITY_4) 
  {
    vpImageMorphology::dilatation<unsigned char>(I, connexity);
  }

  /*!
     \deprecated Provided only for compat with previous releases. Use rather the template function dilatation(vpImage<T> &, const vpConnexityType &)
     
     Add here the description of the function...
   */
  vp_deprecated static void dilatation(vpImage<unsigned char> &I, const vpConnexityType &connexity = CONNEXITY_4)
  {
    vpImageMorphology::dilatation<unsigned char>(I, connexity);
  }
  //@}
#endif
```

**Comment by rolalaro** [comment]
> @rolalaro Thanks, but I'm afraid this PR will break compatibility with previous releases.
> 
> Is there any way of adding non-template functions as deprecated with something like (not tested):
> 
> ```
> #if defined(VISP_BUILD_DEPRECATED_FUNCTIONS)
>   /*!
>     @name Deprecated functions
>   */
>   //@{
>   /*!
>      \deprecated Provided only for compat with previous releases. Use rather the template function erosion(vpImage<T> &, const vpConnexityType &)
>      
>      Add here the description of the function...
>    */
>   vp_deprecated static void erosion(vpImage<unsigned char> &I, const vpConnexityType &connexity = CONNEXITY_4) 
>   {
>     vpImageMorphology::dilatation<unsigned char>(I, connexity);
>   }
> 
>   /*!
>      \deprecated Provided only for compat with previous releases. Use rather the template function dilatation(vpImage<T> &, const vpConnexityType &)
>      
>      Add here the description of the function...
>    */
>   vp_deprecated static void dilatation(vpImage<unsigned char> &I, const vpConnexityType &connexity = CONNEXITY_4)
>   {
>     vpImageMorphology::dilatation<unsigned char>(I, connexity);
>   }
>   //@}
> #endif
> ```

Added in commit [149e26fe](https://github.com/lagadic/visp/pull/1180/commits/149e26fe4b66dc37f92ce19882859f12d19ca339) and tested thanks to additional unit tests"

File: modules/core/include/visp3/core/vpImageMorphology.h [hunk 1]
```diff
@@ -78,15 +78,60 @@ class VISP_EXPORT vpImageMorphology
...
+  #if defined(VISP_BUILD_DEPRECATED_FUNCTIONS)
+  /*!
+    @name Deprecated functions
+  */
+  //@{
+  /*!
+     \deprecated Provided only for compat with previous releases. Use rather the template function erosion(vpImage<T> &, const vpConnexityType &)
+     
+     Add here the description of the function...
+   */
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpImageMorphology.h:118]: Should be updated

File: modules/core/include/visp3/core/vpImageMorphology.h [hunk 2]
```diff
@@ -78,15 +78,60 @@ class VISP_EXPORT vpImageMorphology
...
+  vp_deprecated static void erosion(vpImage<unsigned char> &I, const vpConnexityType &connexity = CONNEXITY_4) 
+  {
+    vpImageMorphology::erosion<unsigned char>(I, connexity);
+  }
+
+  /*!
+     \deprecated Provided only for compat with previous releases. Use rather the template function dilatation(vpImage<T> &, const vpConnexityType &)
+     
+     Add here the description of the function...
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpImageMorphology.h:127]: Should be updated

**Review by fspindle — COMMENTED** [review]
Please update the doc of the 2 deprecated static functions

File: modules/core/include/visp3/core/vpImageMorphology.h (hunk 1 above)
Comment (rolalaro) [reply modules/core/include/visp3/core/vpImageMorphology.h:124]: Sorry, it has been corrected in commit af0e026

File: modules/core/include/visp3/core/vpImageMorphology.h (hunk 2 above)
Comment (rolalaro) [reply modules/core/include/visp3/core/vpImageMorphology.h]: Sorry, it has been corrected in commit af0e026

File: modules/core/test/image/testImageMorphology.cpp [hunk 3]
```diff
@@ -68,10 +71,16 @@ TEST_CASE("Binary image morphology", "[image_morphology]")
...
+      vpImageMorphology::dilatation<unsigned char>(I_morpho, connexity);
 
       CHECK((I_morpho_ref == I_morpho_tpl));
       CHECK((I_morpho_ref == I_morpho));
+
+      #if defined(VISP_BUILD_DEPRECATED_FUNCTIONS)
+      vpImage<unsigned char> I_deprecated_dilatation = I;
+      vpImageMorphology::dilatation(I_deprecated_dilatation, connexity);
+      CHECK((I_morpho_ref == I_deprecated_dilatation));
+      #endif
```
Comment (fspindle) [inline modules/core/test/image/testImageMorphology.cpp:83]: Testing deprecated functions should be removed

File: modules/core/test/image/testImageMorphology.cpp [hunk 4]
```diff
@@ -82,10 +91,16 @@ TEST_CASE("Binary image morphology", "[image_morphology]")
...
     }
```
Comment (fspindle) [inline modules/core/test/image/testImageMorphology.cpp:104]: Same here, remove everything between
```
#if defined(VISP_BUILD_DEPRECATED_FUNCTIONS)
...
#endif
```

File: modules/core/test/image/testImageMorphology.cpp [hunk 5]
```diff
@@ -100,10 +115,16 @@ TEST_CASE("Binary image morphology", "[image_morphology]")
...
+      vpImageMorphology::erosion<unsigned char>(I_morpho, connexity);
 
       CHECK((I_morpho_ref == I_morpho_tpl));
       CHECK((I_morpho_ref == I_morpho));
+
+      #if defined(VISP_BUILD_DEPRECATED_FUNCTIONS)
+      vpImage<unsigned char> I_deprecated_erosion = I;
+      vpImageMorphology::erosion(I_deprecated_erosion, connexity);
+      CHECK((I_morpho_ref == I_deprecated_erosion));
+      #endif
```
Comment (fspindle) [inline modules/core/test/image/testImageMorphology.cpp:127]: Same here

File: modules/core/test/image/testImageMorphology.cpp [hunk 6]
```diff
@@ -115,10 +136,16 @@ TEST_CASE("Binary image morphology", "[image_morphology]")
...
```
Comment (fspindle) [inline modules/core/test/image/testImageMorphology.cpp:148]: Same here

File: modules/core/test/image/testImageMorphology.cpp [hunk 7]
```diff
@@ -137,9 +164,14 @@ TEST_CASE("Binary image morphology", "[image_morphology]")
           1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1};
       vpImage<unsigned char> I_dilatation_ref(image_data_dilatation, 8, 16, true);
       CHECK((I_dilatation_ref == I_dilatation));
+      #if defined(VISP_BUILD_DEPRECATED_FUNCTIONS)
+      vpImage<unsigned char> I_deprecated_dilatation = I;
+      vpImageMorphology::dilatation(I_deprecated_dilatation, connexity);
+      CHECK((I_dilatation_ref == I_deprecated_dilatation));
+      #endif
```
Comment (fspindle) [inline modules/core/test/image/testImageMorphology.cpp:172]: Here also and in other lines below

Comment (rolalaro) [reply modules/core/test/image/testImageMorphology.cpp:143]: Done in commit af0e02632. Thanks for the help =)

File: modules/core/test/image/testImageMorphology.cpp (hunk 6 above)
Comment (rolalaro) [reply modules/core/test/image/testImageMorphology.cpp]: Done in commit af0e02632. Thanks for the help =)

File: modules/core/test/image/testImageMorphology.cpp (hunk 5 above)
Comment (rolalaro) [reply modules/core/test/image/testImageMorphology.cpp]: Done in commit af0e02632. Thanks for the help =)

File: modules/core/test/image/testImageMorphology.cpp (hunk 4 above)
Comment (rolalaro) [reply modules/core/test/image/testImageMorphology.cpp]: Done in commit af0e02632. Thanks for the help =)

File: modules/core/test/image/testImageMorphology.cpp (hunk 3 above)
Comment (rolalaro) [reply modules/core/test/image/testImageMorphology.cpp]: Done in commit af0e02632. Thanks for the help =)

**Pull request merged**